Conch/lib/Conch/Orc/Lifecycle.pm

```diff
@@ -0,0 +1,472 @@
+=head1 NAME
+
+Conch::Orc::Lifecycle
+
+=head1 DESCRIPTION
+
+A Lifecycle is a list of Workflows to execute for a given device role.
+
+=cut
+
+package Conch::Orc::Lifecycle;
+
+use strict;
+use warnings;
+use utf8;
+use v5.20;
+
+use Moo;
+use experimental qw(signatures);
+
+use Try::Tiny;
+use Type::Tiny;
+use Types::Standard qw(Num Str Bool ArrayRef);
+use Types::UUID qw(Uuid);
+use List::MoreUtils qw(uniq);
+
+with "Conch::Role::But";
+with "Moo::Role::ToJSON";
+with "Conch::Role::Timestamps";
+
+use Conch::Time;
+use Conch::Pg;
+use Conch::Orc;
+
+
+=head1 ACCESSORS
+
+=over 4
+
+=item id
+
+UUID. Cannot be written by user.
+
+=cut
+
+has 'id' => (
+	is  => 'rwp',
+	isa => Uuid,
+);
+
+
+=item name
+
+String. Required.
+
+=cut
+
+has 'name' => (
+	is       => 'rw',
+	isa      => Str,
+	required => 1,
+);
+
+
+=item version
+
+Number. Required. Defaults to 1
+
+=cut
+
+has 'version' => (
+	default  => 1,
+	is       => 'rw',
+	isa      => Num,
+	required => 1,
+);
+
+
+
+=item locked
+
+Boolean. Defaults to 0
+
+This governs if steps can be added or removed from the workflow.
+
+=cut
+
+has 'locked' => (
+	is      => 'rw',
+	isa     => Bool,
+	default => 0,
+);
+
+
+=item role_id
+
+UUID. FK'd into C<device_role(id)>. Required.
+
+=cut
+
+has 'role_id' => (
+	is       => 'rw',
+	isa      => Uuid,
+	required => 1,
+);
+
+
+=item plan
+
+Arrayref of Workflow IDs, orderd by the C<plan_order> field. Cannot be written
+by user.
+
+=cut
+
+has 'plan' => (
+	is => 'rwp',
+	isa => ArrayRef,
+	default => sub { [] },
+);
+
+sub _build_serializable_attributes { [qw[
+	id
+	created
+	updated
+	name
+	version
+	locked
+	role_id
+	plan
+]] }
+
+
+
+=head1 METHODS
+
+=head2 from_id
+
+Load a Lifecycle by its UUID
+
+=cut
+
+sub from_id ($class, $id) {
+	my $ret;
+	try {
+		$ret = Conch::Pg->new->db->query(qq|
+			select l.*, array(
+				select wlp.workflow_id
+				from workflow_lifecycle_plan wlp
+				where wlp.lifecycle_id = l.id
+				order by wlp.plan_order
+			) as plan
+			from workflow_lifecycle l
+			where l.deactivated is null and l.id = ?;
+		|, $id)->hash;
+	} catch {
+		Mojo::Exception->throw(__PACKAGE__."->from_id: $_");
+		return undef;
+	};
+
+	return undef unless $ret;
+	return $class->new(_fixup_timestamptzs($ret)->%*);
+}
+
+
+=head2 from_name_latest
+
+Load the latest version of a Lifecycle for a given name
+
+=cut
+
+sub from_name_latest ($class, $name) {
+	my $ret;
+	try {
+		$ret = Conch::Pg->new->db->query(qq|
+			select l.*, array(
+				select wlp.workflow_id
+				from workflow_lifecycle_plan wlp
+				where wlp.lifecycle_id = l.id
+				order by wlp.plan_order
+			) as plan
+			from workflow_lifecycle l
+			where l.deactivated is null and l.id = ?;
+				order by l.version
+				limit 1
```

Comment:
Only one version of a name can be active based on the current schema. This bit is unnecessary unless the schema is changed.